### PR TITLE
[Merged by Bors] - feat: Port/LinearAlgebra.Smodeq

### PR DESCRIPTION
**Pull request by arienmalec**
Depends on `synthInstance.etaExperiment` or permanent replacement.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/LinearAlgebra/Smodeq.lean [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+  (Submodule.Quotient.eq _).2 <|
+    show f (x - y) ∈ V from (f.map_sub x y).symm ▸ (Submodule.Quotient.eq _).1 hxy
+#align smodeq.comap Smodeq.comap
+
+set_option synthInstance.etaExperiment true in
+theorem eval {R : Type _} [CommRing R] {I : Ideal R} {x y : R} (h : x ≡ y [SMOD I]) (f : R[X]) :
+    f.eval x ≡ f.eval y [SMOD I] := by
+  rw [Smodeq.def] at h⊢
+  show Ideal.Quotient.mk I (f.eval x) = Ideal.Quotient.mk I (f.eval y)
+  change Ideal.Quotient.mk I x = Ideal.Quotient.mk I y at h
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Smodeq.lean:117]: See [this zulip thread](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/change.20at/near/341310089)

File: Mathlib.lean [hunk 2]
```diff
@@ -1017,6 +1017,7 @@ import Mathlib.LinearAlgebra.Prod
 import Mathlib.LinearAlgebra.Projection
 import Mathlib.LinearAlgebra.Quotient
 import Mathlib.LinearAlgebra.Ray
+import Mathlib.LinearAlgebra.Smodeq
```
Comment (eric-wieser) [inline Mathlib.lean:1020]:
```suggestion
import Mathlib.LinearAlgebra.SModEq
```

File: Mathlib/LinearAlgebra/Smodeq.lean [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+variable {R : Type _} [Ring R]
+
+variable {M : Type _} [AddCommGroup M] [Module R M] (U U₁ U₂ : Submodule R M)
+
+variable {x x₁ x₂ y y₁ y₂ z z₁ z₂ : M}
+
+variable {N : Type _} [AddCommGroup N] [Module R N] (V V₁ V₂ : Submodule R N)
+
+/-- A predicate saying two elements of a module are equivalent modulo a submodule. -/
+def Smodeq (x y : M) : Prop :=
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Smodeq.lean:32]:
```suggestion
def SModEq (x y : M) : Prop :=
```

Comment (arienmalec) [reply Mathlib/LinearAlgebra/Smodeq.lean]: Done throughout.

File: Mathlib/LinearAlgebra/Smodeq.lean (hunk 1 above)
Comment (arienmalec) [reply Mathlib/LinearAlgebra/Smodeq.lean]: Did the `replace h: <foo> := h` hack.

File: Mathlib/LinearAlgebra/SModEq.lean [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+  (Submodule.Quotient.eq _).2 <|
+    show f (x - y) ∈ V from (f.map_sub x y).symm ▸ (Submodule.Quotient.eq _).1 hxy
+#align smodeq.comap SModEq.comap
+
+set_option synthInstance.etaExperiment true in
+theorem eval {R : Type _} [CommRing R] {I : Ideal R} {x y : R} (h : x ≡ y [SMOD I]) (f : R[X]) :
+    f.eval x ≡ f.eval y [SMOD I] := by
+  rw [SModEq.def] at h⊢
+  show Ideal.Quotient.mk I (f.eval x) = Ideal.Quotient.mk I (f.eval y)
+  replace h: Ideal.Quotient.mk I x = Ideal.Quotient.mk I y := h
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/SModEq.lean:117]:
```suggestion
  replace h : Ideal.Quotient.mk I x = Ideal.Quotient.mk I y := h
```

File: Mathlib/LinearAlgebra/SModEq.lean [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+theorem smul (hxy : x ≡ y [SMOD U]) (c : R) : c • x ≡ c • y [SMOD U] := by
+  rw [SModEq.def] at hxy⊢
+  simp_rw [Quotient.mk_smul, hxy]
+#align smodeq.smul SModEq.smul
+
+theorem zero : x ≡ 0 [SMOD U] ↔ x ∈ U := by rw [SModEq.def, Submodule.Quotient.eq, sub_zero]
+#align smodeq.zero SModEq.zero
+
+set_option synthInstance.etaExperiment true in
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/SModEq.lean:101]: @jcommelin, are you happy with this? Or do you want to try the local deinstancing fix instead?

If we do keep this, then probably we want
```suggestion
set_option synthInstance.etaExperiment true in  -- porting note: workaround for lean4#2074
```
(leanprover/lean4#2074)

**Pull request closed without merging**